driver-core/src/main/com/mongodb/internal/connection/OperationContext.java

```diff
@@ -23,12 +34,87 @@
 public class OperationContext {
     private static final AtomicLong NEXT_ID = new AtomicLong(0);
     private final long id;
+    private final ServerDeprioritization serverDeprioritization;
 
     public OperationContext() {
         id = NEXT_ID.incrementAndGet();
+        serverDeprioritization = new ServerDeprioritization();
     }
 
     public long getId() {
         return id;
     }
+
+    /**
+     * @return The same {@link ServerDeprioritization} if called on the same {@link OperationContext}.
+     */
+    public ServerDeprioritization getServerDeprioritization() {
+        return serverDeprioritization;
+    }
+
+    public static final class ServerDeprioritization {
+        @Nullable
+        private ServerAddress candidate;
```

Comment:
It appears that the asynchronous version may switch threads three times, once in `ConnectionPool` (supported by 'happens-before' in `BlockingQueue`) and again in `AsyncSocketChannel`. Should we mark this field as volatile for safety, or are the existing guarantees sufficient?

Reply:
`ServerDeprioritization` is operation-confined, and an operation is processed sequentially even by the reactive API, regardless of whether multiple threads are involved or not. That "sequential" part must implies the happens-before order between the write and read action pairs that transfer either the whole `ServerDeprioritization` state or part of it across threads.

I think, `TimeoutContext` in the CSOT branch does not use any memory synchronization for its mutable state for the same reason.